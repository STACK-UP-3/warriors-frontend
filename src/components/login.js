import React,{ useEffect } from 'react';
import { useSelector,useDispatch } from 'react-redux';
import M from 'materialize-css';
import { toast } from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';
import { onUserLogin,TokenAuth } from '../redux/actions/loginActions';
import LoginForm from './loginForm';
import Footer from './footer';
import Message from './requestFail';
import SideContainer from './sideContainer';
import SocialAuth from './socialAuth';

toast.configure();

export default ()=>{

    const loginStore = useSelector( store=> store.loginData );
    const dispatch = useDispatch();

    useEffect( ()=>{
        M.AutoInit();
        if(loginStore.error){
            toast.error( <Message />,{
                position: toast.POSITION.TOP_CENTER,
                toastId: 'loginFetchError',
            });
        }
        
        if(loginStore.userInfo.status === 200){
            const token = JSON.stringify(loginStore.userInfo.data.access_token);
            localStorage.setItem('token', token);
            window.location.assign('/dashboard');
        } 

        let token;
        const url  = window.location.search;
        const arr = url.split(""); 
        arr.splice(0,7); 
        
        if(arr.length){
            token = arr.join("")
        }

        if(localStorage.getItem("token")){
            token = localStorage.getItem("token");
        }
        

        if(token) return tokenAuthenticator(token)
    },[]);
    
    const tokenAuthenticator = (token) => { dispatch(TokenAuth(token)); }
    const onFormSubmit = data => { dispatch(onUserLogin(data));  }

    const loginsideContainerTitle = 'Welcome Back to BareFoot Nomad';
    const loginsideContainerSubtitle  = (
            <div>
                <p>
                     If you have not created an account please sign up to use our services
                </p>
                    <br />

                <h4
                    onClick={ ()=> window.location.assign('/signup') }
                >
                    sign up
                </h4>
            </div>
    );
    
    return (
        <div className='loginContainer'>
        <div className='main'>
            <SideContainer title={ loginsideContainerTitle } subtitle ={ loginsideContainerSubtitle  }/>
                
                <div className='formContainer'>

                    <h1> Sign in to BareFoot Nomad </h1>

                        <LoginForm 
                            onSubmit={ onFormSubmit } 
                            userInfo = { loginStore.userInfo } 
                            isLoading = {loginStore.loading}
                            testdata='loginForm' 
                        />
                            <br/>
                        <SocialAuth />
                    
                </div>
        </div>

        <Footer />


    </div>
    );
}
