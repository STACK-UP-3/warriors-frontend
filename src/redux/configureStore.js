import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import promise from 'redux-promise-middleware';
import { composeWithDevTools } from 'redux-devtools-extension';
import reducersCombination from './reducers/index';



export default ()=>{
    const store = createStore(
        reducersCombination, {}, 
        composeWithDevTools(applyMiddleware(thunk,promise))
    );
    
    return store;
}
