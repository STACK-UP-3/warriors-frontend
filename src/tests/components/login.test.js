import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import mockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Login from '../../components/login';
import configureStore from '../../redux/configureStore';
import LoginForm from '../../components/loginForm';
import {tokenData} from '../fixtures/login.testData';

describe('*************** Testing the Login  component ***************',()=>{

    const store = configureStore();
    let onFormSubmit;
    const { location } = window;

    beforeEach(()=>{
        onFormSubmit = jest.fn();
        delete window.location;
        localStorage.removeItem("token");
		window.location = { assign: jest.fn(),search:'' };
    });

    afterAll(() => {
		window.location = location;
        expect(localStorage.getItem).toHaveBeenCalled()
    });
    
    it('Should render the login page correctly', ()=>{
        window.location = { assign: jest.fn(), search : `?token=${tokenData}` };
        const store = mockStore([thunk])({
            loginData: {
                loading: false,
                userInfo: {
                    status: 200,
                    message: 'You have signed in successfully',
                    data:{
                       access_token: tokenData,
                    }
                },
                error: null
            }, 
        });

        const  wrapper = mount(
            <Provider store={ store }> 
                <Login />
            </Provider> 
        );

        const handleSubmit = jest.spyOn(wrapper.find('[testdata="loginForm"]').props(), 'onSubmit');
        wrapper.find('[testdata="loginForm"]').props().onSubmit();
        expect(wrapper).toMatchSnapshot();
        expect(handleSubmit).toHaveBeenLastCalledWith();
    });

    it('should submit upon form call',() => {
       const onSubmitSpy = jest.fn();
       const wrapper = mount(
            <Provider store={ store }> 
                <LoginForm  onSubmit={ onSubmitSpy } userInfo={ { message: 'Test message' } }/>
            </Provider> 
        );

        expect(wrapper).toMatchSnapshot();

        wrapper.find('input').at(0).simulate('change',{ 
                target: { id: 'email', value: 'xxx@yyy.zzz'} 
               });
               
       wrapper.find('input').at(1).simulate('change',{ 
                target: { id: 'password', value: 'kkkkkkkk'} 
               });

            wrapper.find('form').simulate('submit',{
                preventDefault: ()=>{},
            });

        expect(onSubmitSpy).toBeCalled();
    });

    it('Should test form inputs on change and forgot password',() => {
        const testState = { email: 'aaa@bbbb.ccc', password:'jjjjjjj' }
        const wrapper = mount(
             <Provider store={ store }> 
                 <LoginForm  
                  email={ testState.email }
                  password={ testState.password }
                  history = { { push: path => path } }
                  userInfo={ { message: 'Test message' } }
                  />
             </Provider> 
         );

        wrapper.find('input').at(0).simulate('change',{ 
                 target: { id: 'email', value: 'xxx@yyy.zzz'} 
                });
                
        wrapper.find('input').at(1).simulate('change',{ 
                 target: { id: 'password', value: 'kkkkkkkk'} 
                });

        wrapper.find('span').at(0).simulate('click');

        expect(window.location.assign).toHaveBeenCalled();
        
     })

     it('Should test if Fetch does not work',() => {
        const store = mockStore([thunk])({
            loginData: {
                loading: true,
                userInfo: { },
                error: 'TypeError: Failed to fetch',
            }, 
        });

        const wrapper = mount(
             <Provider store={ store }> 
                 <Login />
             </Provider> 
         );

         wrapper.find('h4').simulate('click');

         wrapper.find('form').simulate('submit',{
            preventDefault: ()=>{},
        });

        expect(window.location.assign).toHaveBeenCalled();
        
     })
});
