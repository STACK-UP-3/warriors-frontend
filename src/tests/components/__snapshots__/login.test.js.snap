// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`*************** Testing the Login  component *************** Should render the login page correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <_default>
    <div
      className="loginContainer"
    >
      <div
        className="main"
      >
        <_default
          subtitle={
            <div>
              <p>
                If you have not created an account please sign up to use our services
              </p>
              <br />
              <h4
                onClick={[Function]}
              >
                sign up
              </h4>
            </div>
          }
          title="Welcome Back to BareFoot Nomad"
        >
          <div
            className="sideContainer"
          >
            <div
              className="background"
              style={
                Object {
                  "backgroundImage": "url([object Object])",
                }
              }
            />
            <div
              className="content"
            >
              <h2>
                Welcome Back to BareFoot Nomad
              </h2>
              <div>
                <p>
                  If you have not created an account please sign up to use our services
                </p>
                <br />
                <h4
                  onClick={[Function]}
                >
                  sign up
                </h4>
              </div>
            </div>
          </div>
        </_default>
        <div
          className="formContainer"
        >
          <h1>
             Sign in to BareFoot Nomad 
          </h1>
          <_default
            isLoading={false}
            onSubmit={
              [MockFunction] {
                "calls": Array [
                  Array [],
                ],
                "results": Array [
                  Object {
                    "type": "return",
                    "value": undefined,
                  },
                ],
              }
            }
            testdata="loginForm"
            userInfo={
              Object {
                "data": Object {
                  "access_token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MSwiZnVsbE5hbWUiOiJJUkFEVUtVTkRBIEZpYWNyZSIsImVtYWlsIjoiZmlyYWR1a0B5YWhvby5jb20iLCJyb2xlIjoiUmVxdWVzdGVyIiwiaWF0IjoxNTkwNDAxOTc4LCJleHAiOjE1OTA0MTYzNzh9.K8xbzFSqSghrMzZPZ60EoSd1MylUUxp6NdehPvt9hCA",
                },
                "message": "You have signed in successfully",
                "status": 200,
              }
            }
          >
            <div>
              <form
                className="col s12"
                onSubmit={[Function]}
                testdata="form"
              >
                <div
                  className="form-content"
                >
                  <div
                    className="row"
                    id="textInput"
                  >
                    <div
                      className="input-field col s12"
                    >
                      <input
                        className="validate"
                        disabled={false}
                        id="email"
                        name="Email"
                        onChange={[Function]}
                        testdata="inputChange"
                        type="email"
                      />
                      <label
                        htmlFor="email"
                      >
                        Email
                      </label>
                    </div>
                  </div>
                  <div
                    className="row"
                    id="textInput"
                  >
                    <div
                      className="input-field col s12"
                    >
                      <input
                        className="validate"
                        disabled={false}
                        id="password"
                        name="Password"
                        onChange={[Function]}
                        type="password"
                      />
                      <label
                        htmlFor="password"
                      >
                        Password
                      </label>
                    </div>
                  </div>
                  <div
                    className="contentFooter"
                  >
                    <span
                      id="forgotPassword"
                      onClick={[Function]}
                    >
                      Forgot password?
                    </span>
                    <br />
                     
                    <br />
                    <span
                      className="error"
                    />
                  </div>
                </div>
                 
                <br />
                <div
                  className="form-footer"
                >
                  <div
                    className="submitBtn"
                  >
                    <button
                      className="btn waves-effect waves-light"
                      disabled={true}
                      id="submit"
                    >
                      Login
                    </button>
                  </div>
                </div>
              </form>
            </div>
          </_default>
          <br />
          <SocialAuth>
            <div
              className="socialAuth"
            >
              <p>
                or login with:
              </p>
              <div
                className="socialAuthImg"
              >
                <div
                  id="google"
                  onClick={[Function]}
                >
                  <img
                    src={
                      Object {
                        "process": [Function],
                      }
                    }
                  />
                </div>
                <div
                  id="facebook"
                  onClick={[Function]}
                >
                  <img
                    src={
                      Object {
                        "process": [Function],
                      }
                    }
                  />
                </div>
              </div>
            </div>
          </SocialAuth>
        </div>
      </div>
      <_default>
        <footer
          className="page-footer"
        >
          <div
            className="container"
          >
            <div
              className="footer"
            >
              <span>
                Â© 2020 BareFoot Nomad, All rights reserved.
              </span>
            </div>
          </div>
        </footer>
      </_default>
    </div>
  </_default>
</Provider>
`;

exports[`*************** Testing the Login  component *************** should submit upon form call 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(Symbol.observable): [Function],
    }
  }
>
  <_default
    onSubmit={[MockFunction]}
    userInfo={
      Object {
        "message": "Test message",
      }
    }
  >
    <div>
      <form
        className="col s12"
        onSubmit={[Function]}
        testdata="form"
      >
        <div
          className="form-content"
        >
          <div
            className="row"
            id="textInput"
          >
            <div
              className="input-field col s12"
            >
              <input
                className="validate"
                id="email"
                name="Email"
                onChange={[Function]}
                testdata="inputChange"
                type="email"
              />
              <label
                htmlFor="email"
              >
                Email
              </label>
            </div>
          </div>
          <div
            className="row"
            id="textInput"
          >
            <div
              className="input-field col s12"
            >
              <input
                className="validate"
                id="password"
                name="Password"
                onChange={[Function]}
                type="password"
              />
              <label
                htmlFor="password"
              >
                Password
              </label>
            </div>
          </div>
          <div
            className="contentFooter"
          >
            <span
              id="forgotPassword"
              onClick={[Function]}
            >
              Forgot password?
            </span>
            <br />
             
            <br />
            <span
              className="error"
            >
              Test message
            </span>
          </div>
        </div>
         
        <br />
        <div
          className="form-footer"
        >
          <div
            className="submitBtn"
          >
            <button
              className="btn waves-effect waves-light"
              disabled={true}
              id="submit"
            >
              Login
            </button>
          </div>
        </div>
      </form>
    </div>
  </_default>
</Provider>
`;
